gtsam/discrete/TableFactor.cpp

```diff
@@ -0,0 +1,340 @@
+  for (auto&& key : keys())
+    cout << boost::format(" (%1%,%2%),") % formatter(key) % cardinality(key);
+  cout << " ]" << endl;
+}
+/* ************************************************************************ */
+double TableFactor::operator()(const DiscreteValues& values) const {
+  // Find index of value in O(N) where N is number of keys
+  size_t idx = findIndex(values);
+  return sparse_table_.coeff(idx);
+}
+
+/* ************************************************************************ */
+DecisionTreeFactor TableFactor::operator*(const DecisionTreeFactor& f) const {
+  return toDecisionTreeFactor() * f;
+}
+
+/* ************************************************************************ */
+TableFactor TableFactor::operator*(const TableFactor& f) const {
+  DiscreteKeys keys = unionKeys(f);
+  size_t cardinality = 1;
+  // Create new TableFactor
+  for (DiscreteKey key : keys) cardinality *= key.second;
+  Eigen::SparseVector<double> new_sparse_table_(cardinality);
+  new_sparse_table_.reserve(sparse_table_.nonZeros() *
+                            (cardinality / sparse_table_.size()));
+  TableFactor multiplied_f(keys, new_sparse_table_);
+  for (Eigen::SparseVector<double>::InnerIterator it(f.sparse_table_); it;
+       ++it) {
+    DiscreteValues assignment_f;
+    for (Key key : f.keys_) {
+      assignment_f[key] = f.lazy_cp(key, it.index());
+    }
+    std::vector<DiscreteValues> projections = project(assignment_f);
+    for (DiscreteValues projection : projections) {
+      size_t m_index = multiplied_f.findIndex(projection);
+      multiplied_f.sparse_table_.insert(m_index) =
+          it.value() * operator()(projection);
+    }
+  }
+  return multiplied_f;
+}
+
+/* ************************************************************************ */
+DecisionTreeFactor TableFactor::toDecisionTreeFactor() const {
+  DiscreteKeys dkeys = discreteKeys();
+  std::vector<double> table;
+  for (size_t i = 0; i < sparse_table_.size(); i++) {
+    table.push_back(sparse_table_.coeff(i));
+  }
+  DecisionTreeFactor f(dkeys, table);
+  return f;
+}
+
+/* ************************************************************************ */
+TableFactor::shared_ptr TableFactor::sum(size_t nrFrontals) const {
+  // Create map to keep track of summed values
+  std::map<size_t, double> cs;
+  size_t cardinality = cardinalities_.at(nrFrontals);
+  for (size_t i = 0; i < cardinality; i++) cs[i] = 0;
+  DiscreteKey target_key = discreteKey(nrFrontals);
+  for (Eigen::SparseVector<double>::InnerIterator it(sparse_table_); it; ++it) {
+    cs[lazy_cp(target_key.first, it.index())] += it.value();
+  }
+  std::vector<double> result;
+  for (std::map<size_t, double>::iterator it = cs.begin(); it != cs.end();
+       ++it) {
+    result.push_back(it->second);
+  }
+  return boost::make_shared<TableFactor>(target_key, result);
+}
+
+/* ************************************************************************ */
+TableFactor TableFactor::max(size_t nrFrontals) const {
```

Comment:
can you not call the method below?

Reply:
This function would max by eliminating the variables in incremental order without having to explicitly give an ordering. I thought this was why we had a separate function in DecisionTreeFactor so I added the same feature in TableFactor too. We can call the method below by explicitly giving the ordering in incremental order.